kafka_utils/kafka_rolling_restart/main.py

```diff
@@ -279,6 +312,24 @@ def wait_for_stable_cluster(
         time.sleep(check_interval)
 
 
+def execute_task(checks, host):
+    """execute all the prechecks for the host
+    This trying to run(), and on getting a PrecheckFailedException executes failure().
+    It will try to run run() before exiting(incase of faiilure) to assert if
+    precheck is now valid()
+    """
+    for check in checks:
+        try:
+            check.run(host)
+        except TaskFailedException:
+            check.failure(host)
```

Comment:
Right, but my point is that this can be done in the run() implementation itself. The code executing a task doesn't need to know about the task specific error handling. We can call task.run() then any internal failure can be handled in the task class itself. If eventually, the exception bubbles up then rolling restart will just fail.